detox/src/devices/allocation/drivers/android/emulator/launchEmulatorProcess.js

```diff
@@ -46,7 +47,25 @@ function launchEmulatorProcess(emulatorName, emulatorExec, emulatorLaunchCommand
 
   log = log.child({ child_pid: childProcessPromise.childProcess.pid });
 
-  return childProcessPromise.then(() => true).catch((err) => {
+  // On Android SDK 31 `childProcessPromise` never resolves when Emulator boots,
+  // so we have to continuosly scan for the process output until we detect that the emulator has booted.
+  const monitorOutputWithTimeout = async (timeoutMs) => {
+    const delayMs = 1000;
+    for (let i = 0; i < Math.floor(timeoutMs / delayMs); i++) {
+      await sleep(delayMs);
+      const stdoutContent = fs.readFileSync(tempLog, 'utf8');
+      if (stdoutContent.includes('boot completed')) {
+        return;
+      }
+    }
+    throw new Error('Could not boot emulator');
+  };
+  const monitorOutputWithTimeoutPromise = monitorOutputWithTimeout(
+    3 * 60 * 1000, // 3 mins
+  );
+
+  return Promise.race([childProcessPromise, monitorOutputWithTimeoutPromise])
+  .then(() => true).catch((err) => {
```

Comment:
```suggestion
  const monitorOutputWithTimeout = async (timeoutSeconds) => {
    const SLEEP_INTERVAL = 1000;

    for (let i = 0; i < timeoutSeconds; i++) {
      const stdoutContent = fs.readFileSync(tempLog, 'utf8');
      if (stdoutContent.includes('boot completed')) {
        return;
      }
      
      await sleep(SLEEP_INTERVAL);
    }
    throw new Error('Could not boot emulator');
  };
  const monitorOutputWithTimeoutPromise = monitorOutputWithTimeout(180);

  return Promise.race([childProcessPromise, monitorOutputWithTimeoutPromise])
  .then(() => true).catch((err) => {
```

Reply:
@capezzbr thanks for the contribution! We greatly appreciate this! 👑 

Just my 2 cents here..
1. Why working with ms then dividing it back to seconds? It just adds a redundant complication.
2. Moved the `sleep(...)` after the check, and not before.

Reply:
committed